runtime/src/accounts_db.rs

```diff
@@ -3660,79 +3656,205 @@ impl AccountsDB {
+
+                result
             })
-            .flatten()
             .collect();
-        hashes
+
+        let sum = *overall_sum.lock().unwrap();
+        (hashes, sum)
+    }
+
+    fn de_dup_accounts_from_stores(
+        first_slice: bool,
+        slice: &[CalculateHashIntermediate],
+    ) -> (Vec<Hash>, u128) {
+        let len = slice.len();
+        let mut result: Vec<Hash> = Vec::with_capacity(len);
+
+        let mut sum: u128 = 0;
+        if len > 0 {
+            let mut i = 0;
+            let mut look_for_first_key = !first_slice;
+            'outer: loop {
+                // at start of loop, item at 'i' is the first entry for a given pubkey - unless look_for_first
+                let now = &slice[i];
+                let last = now.pubkey;
+                if !look_for_first_key && now.lamports != ZERO_RAW_LAMPORTS_SENTINEL {
+                    // first entry for this key that starts in our slice
+                    result.push(now.hash);
+                    sum += now.lamports as u128;
+                }
+                for (k, now) in slice.iter().enumerate().take(len).skip(i + 1) {
+                    if now.pubkey != last {
+                        i = k;
+                        look_for_first_key = false;
+                        continue 'outer;
+                    }
+                }
+
+                break; // ran out of items in our slice, so our slice is done
+            }
+        }
+        (result, sum)
+    }
+
+    fn flatten_hashes(hashes: Vec<Vec<Hash>>) -> (Vec<Hash>, Measure, usize) {
+        // flatten vec/vec into 1d vec of hashes in order
+        let mut flat2_time = Measure::start("flat2");
+        let hashes: Vec<Hash> = hashes.into_iter().flatten().collect();
+        flat2_time.stop();
+        let hash_total = hashes.len();
+
+        (hashes, flat2_time, hash_total)
     }
 
+    // input:
+    // vec: unordered, created by parallelism
+    //   vec: [0..bins] - where bins are pubkey ranges
+    //     vec: [..] - items which fin in the containing bin, unordered within this vec
+    // so, assumption is middle vec is bins sorted by pubkey
     fn rest_of_hash_calculation(
-        accounts: (DashMap<Pubkey, CalculateHashIntermediate>, Measure),
+        accounts: (Vec<Vec<CalculateHashIntermediate>>, Measure),
     ) -> (Hash, u64) {
-        let (account_maps, time_scan) = accounts;
+        let (data_sections_by_pubkey, time_scan) = accounts;
 
-        let mut zeros = Measure::start("eliminate zeros");
-        let hashes = Self::remove_zero_balance_accounts(account_maps);
-        zeros.stop();
-        let hash_total = hashes.len();
-        let (ret, (sort_time, hash_time)) =
-            Self::accumulate_account_hashes_and_capitalization(hashes, Slot::default(), false);
+        let (outer, flatten_time, raw_len) =
+            Self::flatten_hash_intermediate(data_sections_by_pubkey);
```

Comment:
Instead of flattening here, I'm curious to see how flattening within Rayon would perform, I've made a suggested change here: https://github.com/carllin/solana/commit/670103db2c8fb9b94440873d6b9404c08a326f0d#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3666

Reply:
I will gladly try this out. I was trying to keep this pr straightforward, so all I did was separate out the flattens into their own area so they can be timed and dealt with in later PRs.

Reply:
another thing to note is that the next PR I am working on is binning the accounts by pubkey value. The result of this is another layer of vectors, which makes the flattening problem worse. For this PR, I was trying to keep standard, easy to understand flattening because the next change was likely to make the complexity higher but the performance better. And I was trying not to over optimize a way point along the route.

I am loving the engagement! It is such a relief that the basic index->non-index code is finally in so we can collaborate on making other parts much faster and better now!

Reply:
I cherry-picked @carllin 's code. Unfortunately, the results are much worse.
accounts_scan 1,082,644us
current pr measurement: 413,866us
I can investigate this more later.

Reply:
no worries, thanks for trying it out!